earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp

```diff
@@ -278,16 +278,105 @@ class StateUpdater::SetStateVisitor : public default_dfs_visitor {
+  private:
+    using RecalcSet = std::unordered_set<SharedString>;
+
+    StateUpdater & updater;
+    DependentStateTree & tree;
+    const AssetDefs::State newState;
+    // This is a shared_ptr because the visitor will be copied several times
+    // during the course of the traversal.
+    const std::shared_ptr<RecalcSet> hasBlockingInputs;
+    const std::shared_ptr<RecalcSet> hasBlockingChildren;
+
+    void SetState(
+        DependentStateTreeVertexDescriptor vertex,
+        AssetDefs::State newState,
+        const WaitingFor & waitingFor) const {
+      SharedString name = tree[vertex].name;
+      AssetDefs::State oldState = tree[vertex].state;
+      if (newState != oldState) {
+        notify(NFY_PROGRESS, "Setting state of '%s' from '%s' to '%s'",
+              name.toString().c_str(), ToString(oldState).c_str(), ToString(newState).c_str());
+        auto version = updater.storageManager->GetMutable(name);
+        if (version) {
+            updater.SetVersionStateAndRunHandlers(version, newState, waitingFor);
+
+          // Get the new state directly from the asset version since it may be
+          // different from the passed-in state
+          auto & data = tree[vertex];
+          data.state = version->state;
+
+          if (AssetDefs::Canceled == newState || AssetDefs::Blocked == newState || AssetDefs::Failed == newState){
+            hasBlockingChildren->insert(version->parents.begin(), version->parents.end());
+            hasBlockingInputs->insert(version->listeners.begin(), version->listeners.end());
+          }
+        }
+        else {
+          // This shoud never happen - we had to successfully load the asset
+          // previously to get it into the tree.
+          notify(NFY_WARN, "Could not load asset '%s' to set state.",
+                name.toString().c_str());
+        }
+      }
+    }
+
+  public:
+    SetBlockingStateVisitor(StateUpdater & updater, DependentStateTree & tree, AssetDefs::State newState) :
+        updater(updater), tree(tree), newState(newState),
+        hasBlockingInputs(std::make_shared<RecalcSet>()),
+        hasBlockingChildren(std::make_shared<RecalcSet>()) {}
+
+    // This function is called after the DFS has completed for every vertex
+    // below this one in the tree. Thus, we don't calculate the state for an
+    // asset until we've calculated the state for all of its inputs and
+    // children.
+    void finish_vertex(
+        DependentStateTreeVertexDescriptor vertex,
+        const DependentStateTree &) const {
+      const AssetVertex & data = tree[vertex];
+      notify(NFY_PROGRESS, "Calculating state for '%s'", data.name.toString().c_str());
+
+      // Set the state for assets in the dependent tree.
+      if (data.inDepTree) {
+        SetState(vertex, newState, {0, 0});
+      }
+      else {
+        // Set the state to blocked if needed, otherwise do nothing
+        auto version = updater.storageManager->Get(data.name);
```

Comment:
This code looks up the version here and inside the SetState function. It would avoid the double lookup if you just pass in the version. Of course, that destroys my previous comment about combining the two SetState functions.